ember-client/shared/src/main/scala/org/http4s/ember/client/internal/WebSocketHelpers.scala

```diff
@@ -16,27 +16,123 @@
 
 package org.http4s.ember.client.internal
 
+import cats.Applicative
 import cats.MonadThrow
 import cats.data.NonEmptyList
+import cats.effect.Async
+import cats.effect.Concurrent
+import cats.effect.MonadCancel
+import cats.effect.Resource
+import cats.effect.SyncIO
+import cats.effect.implicits._
+import cats.effect.std.Queue
 import cats.syntax.all._
+import fs2.io.net.Socket
+import org.http4s.Request
 import org.http4s.Status
 import org.http4s._
+import org.http4s.client.Client
+import org.http4s.client.websocket.WSClient
+import org.http4s.client.websocket.WSConnection
+import org.http4s.client.websocket.WSFrame
 import org.http4s.crypto.Hash
 import org.http4s.crypto.HashAlgorithm
-import org.http4s.headers.Connection
+import org.http4s.ember.core.WebSocketHelpers.decodeFrames
+import org.http4s.ember.core.WebSocketHelpers.frameToBytes
 import org.http4s.headers._
 import org.http4s.websocket.Rfc6455
+import org.http4s.websocket.WebSocketFrame
 import org.typelevel.ci._
+import org.typelevel.vault._
 import scodec.bits.ByteVector
+import org.http4s.client.websocket.WSRequest
 
-private[client] object WebSocketHelpers {
+object WebSocketHelpers {
 
   private[internal] val supportedWebSocketVersion = 13L
 
   private[internal] val upgradeCi = ci"upgrade"
   private[internal] val webSocketProtocol = Protocol(ci"websocket", None)
   private[internal] val connectionUpgrade = Connection(NonEmptyList.of(upgradeCi))
   private[internal] val upgradeWebSocket = Upgrade(webSocketProtocol)
+  private[internal] val exampleSecWebSocketKey = "dGhlIHNhbXBsZSBub25jZQ=="
+
+  // private[this] val clientTranscoder = new FrameTranscoder(true)
+
+  private[internal] def createWebSocketKey[F[_]]: Key[Socket[F]] =
+    Key.newKey[SyncIO, Socket[F]].unsafeRunSync()
+
+  def getSocket[F[_]](client: Client[F], request: Request[F])(implicit
+      F: MonadCancel[F, Throwable]
+  ): Resource[F, Option[Socket[F]]] = {
+    val webSocketKey = createWebSocketKey[F]
+    client
+      .run(request)
+      .flatMap { res =>
+        Resource
+          .eval(validateServerHandshake(res, exampleSecWebSocketKey))
+          .map(isValid => isValid.toOption *> res.attributes.lookup(webSocketKey))
+      }
```

Comment:
- The inner `Resource.eval(ioa).map(a2b)` should be equivalent to a `Resource.eval(ioa.map(a2b))`
- The outer `Resource.flatMap(res => Resource.eval(resToB))` should be equivalent to a `Resource.evalMap`.

Reply:
Thanks will edit this part of code!